src/applicationManager.ts

```diff
@@ -528,3 +546,228 @@ async function getGraphAPIRoles(token: string, graphApiPermissions: string[]) {
   }
   return { appRoleIds, graphAPIObjectId };
 }
+
+/**
+ * Updates an Azure Entra Application, specifically with a collection of new App Roles
+ * @param applicationId The object ID of application registration
+ * @param appRoles A collection (array) of AppRole custom objects to update the application with
+ */
+export async function updateApplicationAppRoles({
+  token,
+  applicationId,
+  appRoles,
+}: {
+  token: string;
+  applicationId: string;
+  appRoles: Array<AppRole>;
+}): Promise<void> {
+  let body = {
+    appRoles: appRoles,
+  };
+  const result = await fetch(
+    `https://graph.microsoft.com/v1.0/applications/${applicationId}`,
+    {
+      method: "PATCH",
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(body),
+    },
+  );
+  await errorHandler("updating application config", result);
+}
+
+/**
+ * Create formatted AppRole object which is expected by graph api
+ * @param id Must be unique compared to other App Roles on this application
+ * @param value Must be unique compared to other App Roles on this application
+ * @return AppRole Returns a promise of structured AppRole type object
+ */
+async function generateStructuredAppRole(
+  displayName: string,
+  description: string,
+  id: string,
+  value: string,
+): Promise<AppRole> {
+  return {
+    allowedMemberTypes: ["User"],
+    description: description,
+    displayName: displayName,
+    id: id,
+    isEnabled: true,
+    value: value,
+  };
+}
+
+/**
+ * Necessary where an app role is deleted, they must all be disabled first. Easier to disable by default and enable when updating.
+ */
+async function disableAppRoles(applicationId: string, token: string) {
```

Comment:
this should only be disabled if it needs to.
we should write some code that checks if it's required, otherwise we're breaking an application while this is running and if the run fails for whatever reason it'll stay broken.

Reply:
Will be looking into this and trying to compare existing app role names with incoming etc and try to only call this when necessary

Reply:
Without disabling the App Roles, yarn test fails with:
```
{"error":{"code":"CannotDeleteOrUpdateEnabledEntitlement","message":"Permission (scope or role) cannot be deleted or updated unless disabled first.
```
This same failure also happens when the application is created for the first time

Looking into if there's a different way around this - but not sure what else to go with

Reply:
https://github.com/hmcts/azure-app-proxy-manager/pull/122/commits/e40e3831eca4e961d4c9d0157fb6bf125eaa0bb7 I believe resolves this @timja  -- it checks existing app role pairs against incoming changes and only disables app roles if there is some deletion taking place